pkg/operator/clustermembercontroller/clustermembercontroller.go

```diff
@@ -90,117 +104,283 @@ func (c *ClusterMemberController) reconcileMembers(ctx context.Context, recorder
 		return fmt.Errorf("unhealthy members found during reconciling members")
 	}
 
-	// etcd is healthy, decide if we need to scale
-	podToAdd, err := c.getEtcdPodToAddToMembership(ctx)
-	switch {
-	case err != nil:
-		return fmt.Errorf("could not get etcd pod: %w", err)
-	case podToAdd == nil:
-		// no more work left to do
-		return nil
+	// Check machine API availability
+	c.isMachineAPIFunctional, err = c.machineAPIChecker.IsFunctional()
+	if err != nil {
+		return fmt.Errorf("failed to determine Machine API availability: %v", err)
 	}
 
-	etcdHost, err := c.getEtcdPeerHostToScale(podToAdd)
+	// Add a learner member if next peer found
+	peerURL, err := c.getEtcdPeerURLToAdd(ctx)
 	if err != nil {
-		return fmt.Errorf("could not get etcd peer host :%w", err)
+		return fmt.Errorf("could not get etcd peerURL to add :%w", err)
 	}
-
-	if machineForHostPendingDeletion, err := c.isMachinePendingDeletionFor(etcdHost); err != nil {
-		return err
-	} else if machineForHostPendingDeletion {
-		return nil
+	if len(peerURL) > 0 {
+		err = c.etcdClient.MemberAddAsLearner(ctx, peerURL)
+		if err != nil {
+			return fmt.Errorf("failed to add learner member :%w", err)
+		}
 	}
 
-	recorder.Eventf("FoundPodToScale", "found pod to add to etcd membership: %v", podToAdd.Name)
-	err = c.etcdClient.MemberAdd(ctx, fmt.Sprintf("https://%s", net.JoinHostPort(etcdHost, "2380")))
-	if err != nil {
-		return fmt.Errorf("could not add member :%w", err)
+	// Attempt to promote a learner member if there is room for more voting members
+	if err := c.ensureEtcdLearnerPromotion(ctx, recorder); err != nil {
+		return fmt.Errorf("failed to promote learner: %w", err)
 	}
+
 	return nil
 }
 
-func (c *ClusterMemberController) isMachinePendingDeletionFor(etcdHost string) (bool, error) {
-	if isFunctional, err := c.machineAPIChecker.IsFunctional(); err != nil {
-		return false, err
-	} else if !isFunctional {
-		// always return false when the machine API is off
-		// otherwise we won't be able to add any member
-		return false, nil
+// getEtcdPeerURLToAdd returns a PeerURL of the next etcd instance to be added as a learner to the cluster. If
+// no eligible member is found return an empty string.
+func (c *ClusterMemberController) getEtcdPeerURLToAdd(ctx context.Context) (string, error) {
+	members, err := c.etcdClient.MemberList(ctx)
```

Comment:
The `etcd-endpoints` configmap only has the list of voting member endpoints, not learners by design.
So checking the current nodes against that list could result in the controller mistakenly trying to add a node that is already a learner in the cluster membership.
```Go
// Ignore if the node's PeerURL is mapped to a member and already part of the quorum.
if isURLMappedToMember(peerURL, members) {
	continue
}
```
> current nodes (which could be also retrieved from the cache)

The nodes are already coming from the node lister which should be using the cache, unless I'm missing something:
```Go
nodes, err := c.masterNodeLister.List(c.masterNodeSelector)
if err != nil {
	return "", err
}
```

Not sure if member list should really be that expensive, although I agree with using the cache as much as possible, but unlikely in this case it seems.

Reply:
> The etcd-endpoints configmap only has the list of voting member endpoints, not learners by design.

yeah, so if all voting members map to all master nodes then we don't have to retrieve the member list.
I think this will be the basic scenario, after all, scaling would be a one-time operation.
does it make sense?